sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -511,6 +510,235 @@ class ConditionalHashJoinIterator(
+              withResource(GpuColumnVector.from(batch)) { builtTable =>
+                withResource(builtTable.filter(trackerTab.getColumn(0))) { filterTab =>
+                  GpuColumnVector.from(filterTab, GpuColumnVector.extractTypes(batch))
+                }
+              }
+            }
+          }
+          // Combine build-side columns with null columns for stream side
+          val ret = withResource(filteredBatch) { builtBatch =>
+            val numFilterRows = builtBatch.numRows()
+            if (numFilterRows > 0) {
+              val streamColumns = streamAttributes.safeMap { attr =>
+                GpuColumnVector.fromNull(numFilterRows, attr.dataType)
+              }
+              withResource(new ColumnarBatch(streamColumns.toArray, numFilterRows)) { streamBatch =>
+                buildSide match {
+                  case GpuBuildRight =>
+                    Some(GpuColumnVector.combineColumns(streamBatch, builtBatch))
+                  case GpuBuildLeft =>
+                    Some(GpuColumnVector.combineColumns(builtBatch, streamBatch))
+                }
+              }
+            } else {
+              None
+            }
+          }
+          builtSideTracker = None
+          ret
+        }
+      }
+    }
+  }
+
+  private def trueColumnTable(numRows: Int): Table = {
+    withResource(Scalar.fromBool(true)) { trueScalar =>
+      withResource(ai.rapids.cudf.ColumnVector.fromScalar(trueScalar, numRows)) {
+        new Table(_)
+      }
+    }
+  }
+
+  // Create a boolean column to indicate which gather map rows are valid.
+  private def validIndexMask(gatherView: ColumnView): ColumnView = {
+    withResource(Scalar.fromInt(Int.MinValue)) { invalidIndex =>
+      gatherView.notEqualTo(invalidIndex)
+    }
+  }
+
+  /**
+   * Update the tracking mask for the build side.
+   */
+  private def updateTrackingMask(buildSideGatherMap: LazySpillableGatherMap): Unit = {
+    // Filter the build side gather map to remove invalid indices
+    val numGatherMapRows = buildSideGatherMap.getRowCount.toInt
+    val filteredGatherMap = {
+      withResource(buildSideGatherMap.toColumnView(0, numGatherMapRows)) { gatherView =>
+        withResource(gatherView.copyToColumnVector()) { gatherVec =>
+          withResource(new Table(gatherVec)) { gatherTab =>
+            withResource(validIndexMask(gatherView)) { mask =>
+              gatherTab.filter(mask)
+            }
+          }
+        }
+      }
+    }
+    // Get the current tracking table, or all true table to start with
+    val builtTrackingTable = builtSideTracker.map { spillableBatch =>
+      withResource(spillableBatch.releaseBatch()) { trackingBatch =>
+        GpuColumnVector.from(trackingBatch)
+      }
+    }.getOrElse {
+      trueColumnTable(numBuiltRows)
+    }
+    // Update all hits in the gather map with false (no longer needed) in the tracking table
+    val updatedTrackingTable = {
+      withResource(filteredGatherMap) { filteredMap =>
```

Comment:
We leak filteredGatherMap if something above throws (e.g.: spillableBatch.releaseBatch or trueColumnTable)

Reply:
Good catch.  Thanks!